src/psbt/finalizer.rs

```diff
@@ -0,0 +1,371 @@
+    if let Some(ref witness_utxo) = inp.witness_utxo {
+        script_pubkey = &witness_utxo.script_pubkey;
+    } else if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
+        let vout = psbt.global.unsigned_tx.input[index].previous_output.vout;
+        script_pubkey = &non_witness_utxo
+            .output
+            .get(vout as usize)
+            .expect("Invalid Input index")
+            .script_pubkey;
+    } else {
+        return Err(InputError::MissingUtxo);
+    }
+    Ok(script_pubkey)
+}
+
+// Get the amount being spent for the psbt input
+fn get_amt(psbt: &Psbt, index: usize) -> Result<u64, InputError> {
+    let amt;
+    let inp = psbt.inputs.get(index).expect("Invalid Input index");
+    if let Some(ref witness_utxo) = inp.witness_utxo {
+        amt = witness_utxo.value;
+    } else if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
+        let vout = psbt.global.unsigned_tx.input[index].previous_output.vout;
+        amt = non_witness_utxo
+            .output
+            .get(vout as usize)
+            .expect("Invalid Input index")
+            .value;
+    } else {
+        return Err(InputError::MissingUtxo);
+    }
+    Ok(amt)
+}
+
+// Create a descriptor from unfinalized PSBT input.
+// Panics on out of bound input index for psbt
+// Also sanity checks that the witness script and
+// redeem script are consistent with the script pubkey.
+// Does *not* check signatures
+fn get_descriptor(psbt: &Psbt, index: usize) -> Result<Descriptor<PublicKey>, InputError> {
+    // Figure out Scriptpubkey
+    let script_pubkey = get_scriptpubkey(psbt, index)?;
+    let inp = psbt.inputs.get(index).expect("Invalid Input index");
+    // 1. `PK`: creates a `Pk` descriptor(does not check if partial sig is given)
+    if script_pubkey.is_p2pk() {
+        let script_pubkey_len = script_pubkey.len();
+        let pk_bytes = &script_pubkey.to_bytes();
+        match bitcoin::PublicKey::from_slice(&pk_bytes[1..script_pubkey_len - 1]) {
+            Ok(pk) => return Ok(Descriptor::Pk(pk)),
+            Err(e) => return Err(InputError::from(e)),
+        }
+    } else if script_pubkey.is_p2pkh() {
+        // 1. `Pkh`: creates a `PkH` descriptor if partial_sigs has the correponding pk
+        let partial_sig_contains_pk = inp
+            .partial_sigs
+            .iter()
+            .filter(|&(&pk, _sig)| {
+                // Indirect way to check the equivalence of pubkey-hashes.
+                // Create a pubkey hash and check if they are the same.
+                let addr = bitcoin::Address::p2pkh(&pk, bitcoin::Network::Bitcoin);
+                *script_pubkey == addr.script_pubkey()
+            })
+            .next();
+        match partial_sig_contains_pk {
+            Some((pk, _sig)) => {
+                return Ok(Descriptor::Pkh(pk.to_owned()));
+            }
+            None => {
+                return Err(InputError::MissingPubkey);
+            }
```

Comment:
more idiomatic (and readable) to move the `return` out of the match so it is clear there are no non-returning branches
```
return match partial_sig_contains_pk {
    Some((pk, _sig)) => Ok(Descriptor::Pkh(pk.to_owned())),
    None => Err(InputError::MissingPubkey),
};
```

Reply:
also every single `return` in this function could be removed since they appear at the end of the block (you need to change a couple `if`s to `else if`s). This would also be clearer since it emphasizes that this giant if statement is the only thing going on.

Reply:
Removing all returns as it is the last statement

Reply:
Rather, should I keep the short circuit return paths? Maybe those are more clear

Reply:
I only saw one that I'd consider a "short-circuit return path", which is the one I complained about below because it was immediately followed by an `.unwrap()`.